src/main/kotlin/world/trait/Trait.kt

```diff
@@ -25,45 +28,125 @@ sealed class Trait(val name: String) {
 
     @Serializable @JvmInline
     @SerialName("TraitID")
-    value class ID(val id: String) { override fun toString() = id }
+    value class ID(val id: String) {
+        override fun toString() = id
+        inline fun trait() = Yegg.world.traits[this]
+    }
 
     val id = ID(NanoID.newID())
     val vTrait = VTrait(id)
 
-    val traits: MutableList<ID> = mutableListOf()
+    // Traits we inherit from.
+    val parents: MutableList<ID> = mutableListOf()
+    // Traits which inherit from us.
+    val children: MutableList<ID> = mutableListOf()
 
     val commands: MutableSet<Command> = mutableSetOf()
     val verbs: MutableMap<String, Verb> = mutableMapOf()
-    val props: MutableMap<String, Value> = mutableMapOf()
 
+    // All props we define directly OR inherit.
+    val props: MutableMap<String, Propval> = mutableMapOf()
+
+    // All objects which inherit this trait (directly or transitively).
```

Comment:
I'm sure you thought about this, but it seems like keeping track of descendants this way will lead to some very large sets of descendants at the top of the tree.  Would it be better to only track direct descendants in each trait, and then traverse the child traits?

Reply:
I did think about it, but that was before I added $root to be honest.  I think you may be right about this.  I only need to look at $trait.objects when we add/remove properties; since we track $trait.children and .parents, we can traverse that to get to all the inheritors when we do those things.  And those ops don't need to be "hot path optimal".

I believe I will make this change.